feat: add `rg.list_datasets` for API v0 datasets

alvarobartt opened the pull request:
# Description

This PR adds the `list_datasets` method via `import argilla as rg; rg.list_datasets` to list the `TextClassification`, `TokenClassification`, and `Text2Text` datasets via calling the endpoint `GET /api/datasets` in the API v0.

Closes #3250 

**Type of change**

- [X] New feature (non-breaking change which adds functionality)

**How Has This Been Tested**

- [X] Add integration tests for `list_datasets` in the Python SDK

**Checklist**

- [ ] I added relevant documentation
- [X] follows the style guidelines of this project
- [X] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [X] I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

alvarobartt committed:
feat: add `list_datasets` from API v0

alvarobartt committed:
fix(style): remove unused import

alvarobartt committed:
test: add `test_list_datasets`

alvarobartt committed:
docs: update `CHANGELOG.md`

gabrielmbmb commented on src/argilla/client/sdk/datasets/api.py:
```diff
@@ -38,6 +38,28 @@ def get_dataset(client: AuthenticatedClient, name: str) -> Response[Dataset]:
+            content=response.content,
+            headers=response.headers,
+            parsed=parsed_response,
+        )
```
```suggestion
        response_obj = Response.from_httpx_response(response)
        response_obj.parsed = [Dataset(**dataset) for dataset in response.json()]
        return response_obj
```

alvarobartt replied:
Makes sense, didn't do it because this is low priority and changing something would imply changing the whole file (if not every single file that it's outdated)

gabrielmbmb commented on src/argilla/client/sdk/datasets/api.py:
```diff
@@ -38,6 +38,28 @@ def get_dataset(client: AuthenticatedClient, name: str) -> Response[Dataset]:
 
 
+def list_datasets(client: AuthenticatedClient, workspace: Optional[str] = None) -> Response[List[Dataset]]:
+    url = "{}/api/datasets".format(client.base_url)
```
We could try to update all the functions to use the f-string

frascuchon commented on src/argilla/client/sdk/datasets/api.py:
```diff
@@ -38,6 +38,28 @@ def get_dataset(client: AuthenticatedClient, name: str) -> Response[Dataset]:
+        headers=client.get_headers(),
+        cookies=client.get_cookies(),
+        timeout=client.get_timeout(),
+    )
```
If the `rg.set_workspace` function is launched at some point. the `client.get_headers()` will contain the workspace header `"X-Argilla-Workspace"` which will be used to filter datasets. If you need to list all dataset available to user, you should remove the header in this call, and only keep the workspace logic through the query params.

```suggestion
    headers = client.get_headers().copy()
    headers.pop(_constans.WORKSPACE_HEADER_NAME, None)
    
    response = httpx.get(
        url=url,
        params={"workspace": workspace} if workspace else None,
        headers=headers,
        cookies=client.get_cookies(),
        timeout=client.get_timeout(),
    )
```

frascuchon commented on tests/integration/client/sdk/api/test_datasets.py:
```diff
@@ -72,3 +72,25 @@ def test_build_response(status_code, expected):
+
+@pytest.mark.parametrize("role", [UserRole.admin, UserRole.owner])
+@pytest.mark.asyncio
+async def test_list_datasets(role: UserRole) -> None:
```
Please, add a test that covers fetching datasets from 2 different workspaces.

alvarobartt committed:
fix: use `f-string` & `from_httpx_response`

Co-authored-by: Gabriel Martin <gabriel@argilla.io>

alvarobartt committed:
fix: return-type hints

alvarobartt committed:
fix: set `X-Argilla-Workspace` header to `None`

It avoids an user with an activated workspace to list all the available workspaces, as just that workspace will be listed.

Co-authored-by: Paco Aranda <francis@argilla.io>

alvarobartt committed:
test: add `test_list_datasets_from_multiple_workspaces`

frascuchon commented on src/argilla/client/sdk/datasets/api.py:
```diff
@@ -35,11 +36,40 @@ def get_dataset(client: AuthenticatedClient, name: str) -> Response[Dataset]:
+        response_obj = Response.from_httpx_response(response)
+        response_obj.parsed = Dataset(**response.json())
+        return response_obj
+    return handle_response_error(response)
```
Indeed, this line will raise an error instead of returning something. 
```suggestion
    handle_response_error(response)
```

frascuchon commented on src/argilla/client/sdk/datasets/api.py:
```diff
@@ -14,19 +14,20 @@
 @lru_cache(maxsize=None)
-def get_dataset(client: AuthenticatedClient, name: str) -> Response[Dataset]:
-    url = "{}/api/datasets/{name}".format(client.base_url, name=name)
+def get_dataset(client: AuthenticatedClient, name: str) -> Response[Union[Dataset, ErrorMessage, HTTPValidationError]]:
```
Errors occurred here will be raised instead of returned as part of a response.
```suggestion
def get_dataset(client: AuthenticatedClient, name: str) -> Response[Dataset]:
```

frascuchon commented on src/argilla/client/sdk/datasets/api.py:
```diff
@@ -35,11 +36,40 @@ def get_dataset(client: AuthenticatedClient, name: str) -> Response[Dataset]:
+        response_obj = Response.from_httpx_response(response)
+        response_obj.parsed = [Dataset(**dataset) for dataset in response.json()]
+        return response_obj
+    return handle_response_error(response)
```
```suggestion
    handle_response_error(response)
```

frascuchon commented on src/argilla/client/sdk/datasets/api.py:
```diff
@@ -35,11 +36,40 @@ def get_dataset(client: AuthenticatedClient, name: str) -> Response[Dataset]:
+
+def list_datasets(
+    client: AuthenticatedClient, workspace: Optional[str] = None
+) -> Response[Union[List[Dataset], ErrorMessage, HTTPValidationError]]:
```
```suggestion
) -> Response[List[Dataset]]:
```

frascuchon commented on src/argilla/client/sdk/datasets/api.py:
```diff
@@ -49,11 +79,15 @@ def copy_dataset(client: AuthenticatedClient, name: str, json_body: CopyDatasetR
 
-def delete_dataset(client: AuthenticatedClient, name: str) -> httpx.Response:
-    url = "{}/api/datasets/{name}".format(client.base_url, name=name)
+def delete_dataset(client: AuthenticatedClient, name: str) -> Response[Union[ErrorMessage, HTTPValidationError]]:
```
```suggestion
def delete_dataset(client: AuthenticatedClient, name: str) -> Response[dict]:
```

frascuchon commented on src/argilla/client/sdk/datasets/api.py:
```diff
@@ -49,11 +79,15 @@ def copy_dataset(client: AuthenticatedClient, name: str, json_body: CopyDatasetR
+        response_obj = Response.from_httpx_response(response)
+        response_obj.parsed = Dataset(**response.json())
+        return response_obj
+    return handle_response_error(response)
```
```suggestion
    handle_response_error(response)
```

frascuchon commented on src/argilla/client/sdk/datasets/api.py:
```diff
@@ -35,11 +36,40 @@ def get_dataset(client: AuthenticatedClient, name: str) -> Response[Dataset]:
-    url = "{}/api/datasets/{name}:copy".format(client.base_url, name=name)
+def copy_dataset(
+    client: AuthenticatedClient, name: str, json_body: CopyDatasetRequest
+) -> Response[Union[Dataset, ErrorMessage, HTTPValidationError]]:
```
```suggestion
) -> Response[Dataset]:
```

alvarobartt committed:
fix: skip `return` from `handle_response_errors` and fix type-hints

Co-authored-by: Paco Aranda <francis@argilla.io>
